solidity/contracts/token-distribution/TokenDistributor.sol

```diff
@@ -0,0 +1,251 @@
+/// based on their address on some chain (A), and operates under a different address
+/// on the host chain (B) where tokens distribution takes place (e.g. a recipient
+/// is as a multi-sig wallet on chain A and receives tokens assignment on chain B,
+/// they have to sign an address on chain B they control with a signature calculated
+/// with a private key controlling the address on chain A).
+/// @dev This contract is based on the Uniswap's Merkle Distributor
+/// https://github.com/Uniswap/merkle-distributor with some modifications:
+/// - added 'allocate' function that will be called to allocate tokens for a
+///   merkle root,
+/// - added a possibility for a recipient to redirect tokens withdrawal to another
+///   address by providing a signature over that address,
+/// - added a possibility for the owner to recover unclaimed tokens after the
+///   unclaimed unlock duration that is configurable on tokens allocation.
+contract TokenDistributor is Ownable {
+    using SafeERC20 for IERC20;
+
+    // Distributed token.
+    IERC20 public token;
+
+    // Merkle tree root.
+    bytes32 public merkleRoot;
+
+    // Timestamp after which allocated tokens can be recovered from the contract
+    // by the owner. If the value is zero the recovery is not possible.
+    uint256 public unclaimedUnlockTimestamp;
+
+    mapping(uint256 => uint256) private claimedBitMap;
+
+    event TokensAllocated(bytes32 merkleRoot, uint256 amount);
+
+    event TokensClaimed(
+        uint256 indexed index,
+        address indexed recipient,
+        address indexed destination,
+        uint256 amount
+    );
+
+    event TokensRecovered(address destination, uint256 amount);
+
+    constructor(IERC20 _token) public {
+        token = _token;
+    }
+
+    /// @notice Claim assigned tokens. The function can be used to withdraw tokens
+    /// assigned to the recipient in the merkle tree distribution. The caller has
+    /// to provide merkle data: recipient address, index, amount, and merkle proof.
+    /// Anyone can call this function. The function requires a destination address
+    /// for a transfer to be provided. The destination address should be signed
+    /// by the recipient. To construct the message to signing recipient has to
+    /// hash destination address with keccak256 and prefix the obtained digest with
+    /// Ethereum specific `\x19Ethereum Signed Message:\n32`. When using web3's
+    /// signing function the prefixing is done automatically, so the signing operation
+    /// can look like: `sign(keccak256(destination), privateKey)`.
+    /// The signature should be provided with the call to this function. This is
+    /// to confirm that the recipient approved the destination address for a transfer.
+    /// The solution allows cross-chain allocations, where a recipient from a
+    /// different chain does not operate under the same address on the host chain
```

Comment:
Here also I'd drop `host` and say just `... on the chain where tokens were allocated ...`